File: core 2/src/main/java/com/dev/core/service/TimeLogService.java
package com.dev.core.service;


import java.time.LocalDate;
import java.util.List;
import java.util.Map;

import com.dev.core.model.TimeLogDTO;

public interface TimeLogService {

    // ---------------------------------------------------------
    // TIMER OPERATIONS
    // ---------------------------------------------------------

    /**
     * Starts a new timer for a user.
     * Task or Bug is optional. Project will be auto-derived.
     */
    TimeLogDTO startTimer(Long userId, Long taskId, Long bugId, String note);

    /**
     * Stops the currently active timer for the user.
     */
    TimeLogDTO stopTimer(Long userId);

    /**
     * Gets the active timer entry for a user, if present.
     */
    TimeLogDTO getActiveTimer(Long userId);

    // ---------------------------------------------------------
    // CRUD OPERATIONS (manual time log entries)
    // ---------------------------------------------------------

    /**
     * Create a manual time entry (not timer-based).
     */
    TimeLogDTO createManualEntry(TimeLogDTO dto);

    /**
     * Update an existing time entry manually.
     */
    TimeLogDTO updateManualEntry(Long id, TimeLogDTO dto);

    /**
     * Delete a time entry.
     */
    void delete(Long id);

    // ---------------------------------------------------------
    // FETCH OPERATIONS
    // ---------------------------------------------------------

    /**
     * Get all time logs for a user (optionally filtered).
     */
    List<TimeLogDTO> getTimeLogs(
            Long userId,
            Long projectId,
            Long taskId,
            Long bugId,
            LocalDate fromDate,
            LocalDate toDate
    );

    /**
     * Get time logs for a specific day.
     */
    List<TimeLogDTO> getDailyLogs(Long userId, LocalDate date);

    /**
     * Get logs in a week (Mon–Sun).
     */
    List<TimeLogDTO> getWeeklyLogs(Long userId, LocalDate weekStartDate);

    /**
     * Get logs for a whole month.
     */
    List<TimeLogDTO> getMonthlyLogs(Long userId, int year, int month);

    /**
     * Get total time spent by user for a specific day.
     */
    Long getDailyTotalMinutes(Long userId, LocalDate date);

    /**
     * Get total time spent per day in a week.
     */
    Map<LocalDate, Long> getWeeklyBreakdown(Long userId, LocalDate weekStartDate);

    /**
     * Get total time spent per day in a month.
     */
    Map<LocalDate, Long> getMonthlyBreakdown(Long userId, int year, int month);

    /**
     * Total time spent on a project by a user.
     */
    Long getTotalTimeForProject(Long userId, Long projectId);

    /**
     * Total time spent on a task.
     */
    Long getTotalTimeForTask(Long userId, Long taskId);

    /**
     * Total time spent on a bug.
     */
    Long getTotalTimeForBug(Long userId, Long bugId);

    // ---------------------------------------------------------
    // COMPANY-LEVEL ANALYTICS
    // ---------------------------------------------------------

    /**
     * Total time logged by all users for a specific day.
     */
    Long getCompanyDailyTotal(LocalDate date);

    /**
     * Total time logged by all users for a week.
     */
    Long getCompanyWeeklyTotal(LocalDate weekStart);

    /**
     * Total time logged by all users in a month.
     */
    Long getCompanyMonthlyTotal(int year, int month);

    /**
     * Breakdown of total time per user for a day.
     */
    Map<Long, Long> getCompanyDailyBreakdown(LocalDate date);

    /**
     * Breakdown of time per project for a user.
     */
    Map<Long, Long> getUserProjectBreakdown(Long userId, LocalDate from, LocalDate to);

    /**
     * Breakdown of time per project (company level).
     */
    Map<Long, Long> getCompanyProjectBreakdown(LocalDate from, LocalDate to);

    // ---------------------------------------------------------
    // REPORTING UTILITIES
    // ---------------------------------------------------------

    /**
     * Get time summary for UI dashboard calendar view.
     */
    Map<LocalDate, Long> getCalendarSummary(Long userId, int year, int month);
}

File: CORE-FE/src/services/timelog.service.ts
import axiosInstance from '../axiosInstance';
import type { TimeLogDTO, CreateTimeLogDTO, TimeBreakdown, ProjectBreakdown, CalendarSummary } from '../types/timelog.types';

const TIMELOG_API_BASE = '/timelogs';

export const timelogService = {
    // --------------------------------------------------------------
    // TIMER OPERATIONS
    // --------------------------------------------------------------
    startTimer: async (userId: number, taskId?: number, bugId?: number, note?: string): Promise<TimeLogDTO> => {
        const response = await axiosInstance.post(`${TIMELOG_API_BASE}/start`, null, {
            params: { userId, taskId, bugId, note }
        });
        return response.data.data;
    },

    stopTimer: async (userId: number): Promise<TimeLogDTO> => {
        const response = await axiosInstance.post(`${TIMELOG_API_BASE}/stop`, null, {
            params: { userId }
        });
        return response.data.data;
    },

    getActiveTimer: async (userId: number): Promise<TimeLogDTO | null> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/active`, {
            params: { userId }
        });
        return response.data.data;
    },

    // --------------------------------------------------------------
    // MANUAL ENTRY
    // --------------------------------------------------------------
    createManualEntry: async (dto: CreateTimeLogDTO): Promise<TimeLogDTO> => {
        const response = await axiosInstance.post(TIMELOG_API_BASE, dto);
        return response.data.data;
    },

    updateManualEntry: async (id: number, dto: Partial<TimeLogDTO>): Promise<TimeLogDTO> => {
        const response = await axiosInstance.put(`${TIMELOG_API_BASE}/${id}`, dto);
        return response.data.data;
    },

    deleteEntry: async (id: number): Promise<void> => {
        await axiosInstance.delete(`${TIMELOG_API_BASE}/${id}`);
    },

    // --------------------------------------------------------------
    // FETCH LOGS
    // --------------------------------------------------------------
    getTimeLogs: async (
        userId: number,
        projectId?: number,
        taskId?: number,
        bugId?: number,
        fromDate?: string,
        toDate?: string
    ): Promise<TimeLogDTO[]> => {
        const response = await axiosInstance.get(TIMELOG_API_BASE, {
            params: { userId, projectId, taskId, bugId, fromDate, toDate }
        });
        return response.data.data;
    },

    getDailyLogs: async (userId: number, date: string): Promise<TimeLogDTO[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/daily`, {
            params: { userId, date }
        });
        return response.data.data;
    },

    getWeeklyLogs: async (userId: number, weekStart: string): Promise<TimeLogDTO[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/weekly`, {
            params: { userId, weekStart }
        });
        return response.data.data;
    },

    getMonthlyLogs: async (userId: number, year: number, month: number): Promise<TimeLogDTO[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/monthly`, {
            params: { userId, year, month }
        });
        return response.data.data;
    },

    // --------------------------------------------------------------
    // TOTALS
    // --------------------------------------------------------------
    getDailyTotal: async (userId: number, date: string): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/daily/total`, {
            params: { userId, date }
        });
        return response.data.data;
    },

    getWeeklyBreakdown: async (userId: number, weekStart: string): Promise<TimeBreakdown[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/weekly/breakdown`, {
            params: { userId, weekStart }
        });
        return response.data.data;
    },

    getMonthlyBreakdown: async (userId: number, year: number, month: number): Promise<TimeBreakdown[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/monthly/breakdown`, {
            params: { userId, year, month }
        });
        return response.data.data;
    },

    getTotalTimeForProject: async (userId: number, projectId: number): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/project/total`, {
            params: { userId, projectId }
        });
        return response.data.data;
    },

    getTotalTimeForTask: async (userId: number, taskId: number): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/task/total`, {
            params: { userId, taskId }
        });
        return response.data.data;
    },

    getTotalTimeForBug: async (userId: number, bugId: number): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/bug/total`, {
            params: { userId, bugId }
        });
        return response.data.data;
    },

    // --------------------------------------------------------------
    // COMPANY ANALYTICS
    // --------------------------------------------------------------
    getCompanyDailyTotal: async (date: string): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/company/daily-total`, {
            params: { date }
        });
        return response.data.data;
    },

    getCompanyWeeklyTotal: async (weekStart: string): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/company/weekly-total`, {
            params: { weekStart }
        });
        return response.data.data;
    },

    getCompanyMonthlyTotal: async (year: number, month: number): Promise<number> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/company/monthly-total`, {
            params: { year, month }
        });
        return response.data.data;
    },

    getCompanyDailyBreakdown: async (date: string): Promise<any> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/company/daily-breakdown`, {
            params: { date }
        });
        return response.data.data;
    },

    getUserProjectBreakdown: async (userId: number, fromDate: string, toDate: string): Promise<ProjectBreakdown[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/user/project-breakdown`, {
            params: { userId, fromDate, toDate }
        });
        return response.data.data;
    },

    getCompanyProjectBreakdown: async (fromDate: string, toDate: string): Promise<ProjectBreakdown[]> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/company/project-breakdown`, {
            params: { fromDate, toDate }
        });
        return response.data.data;
    },

    // --------------------------------------------------------------
    // CALENDAR SUMMARY
    // --------------------------------------------------------------
    getCalendarSummary: async (userId: number, year: number, month: number): Promise<CalendarSummary> => {
        const response = await axiosInstance.get(`${TIMELOG_API_BASE}/calendar`, {
            params: { userId, year, month }
        });
        return response.data.data;
    }
};

export default timelogService;
